### ISSUE TITLE
More idiomatic map support for tech.ml.dataset.impl.dataset.DataSet

### ISSUE THREAD

**Issue by joinr** [techascent/tech.ml.dataset#108]
This fleshes out the tech.ml.dataset.impl.dataset.DataSet
    type with support for more of the clojure conveniences
    and the ability to view it as map of {column-name column-value}.

    Additions include consistent java.util.Map and clojure
    IPersistentMap support.  Equality is now supported via
    hash and hasheq.

    Changes to legacy behavior are primarily on the
    iterator and function invocation implementations.

    The iterator provided (and still does) a sequence of
    columns.  This is inconsistent with the seq idiom
    for maps, which projects the map onto a sequence
    of map entries, or [k v] pairs.  Similarly,
    map conj behavior allows for conjoining [k v]
    pairs or map entries.

    To be consistent with clojure, we retain the iterator
    as is, but provide a seq implementation that
    conforms to the expected map projection.

    The extant IFn implementation behaved like a
    map for the 2-arg arity, providing a lookup
    facility.  However, the 3-arg variant serves
    as a replacement for assoc to update the dataset
    column name and column.  This is inconsistent
    with clojure idioms.  A new implementation
    is provided that follows the clojure idiom
    of lookup with an optional default value,
    so the IFn implementation is now consistent
    with the clojure idiom for clojure.core/get.

    Notably, the IFn change and the change in seq
    implementation are possibly backwards incompatible
    depending on expectations in legacy code.

[patch.diff](https://github.com/techascent/tech.ml.dataset/files/4890555/patch.diff.txt)

```clojure
user>(require '[tech.ml.dataset :as ds])

user> (def csv-data (ds/->dataset "https://github.com/techascent/tech.ml.dataset/raw/master/test/data/stocks.csv"))
#'user/csv-data
user> (def other-csv-data (ds/->dataset "https://github.com/techascent/tech.ml.dataset/raw/master/test/data/stocks.csv"))
#'user/other-csv-data

;;checking equality
user> (= csv-data {:a 2})
false
user> (= csv-data other-csv-data)
true
user> (= csv-data :hello)
false
user> (def csv-map (into {} (seq csv-data)))
#'user/csv-map
user> (keys csv-map)
("symbol" "date" "price")
user> (= csv-data csv-map)
true
user> (keys csv-data)
("symbol" "date" "price")

;;messing with map access
user> (->> csv-data vals first (take 10))
("MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT")

user> (for [[k col] csv-data] [k (take 10 col)])
(["symbol"
  ("MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT")]
 ["date"
  (#object[java.time.LocalDate 0x58214e62 "2000-01-01"]
   #object[java.time.LocalDate 0x2187ece9 "2000-02-01"]
   #object[java.time.LocalDate 0x10773c8 "2000-03-01"]
   #object[java.time.LocalDate 0x4ea1f266 "2000-04-01"]
   #object[java.time.LocalDate 0xf05b55c "2000-05-01"]
   #object[java.time.LocalDate 0x117bf0f1 "2000-06-01"]
   #object[java.time.LocalDate 0x2b584825 "2000-07-01"]
   #object[java.time.LocalDate 0x6a2bbdea "2000-08-01"]
   #object[java.time.LocalDate 0x48a64aa6 "2000-09-01"]
   #object[java.time.LocalDate 0x614c73b "2000-10-01"])]
 ["price" (39.81 36.35 43.22 28.37 25.45 32.54 28.4 28.4 24.53 28.02)])

;;Changed FN invocation...

;;This acts as before...
user> (csv-data (first (keys csv-data)))
#tech.ml.dataset.column<string>[560]
symbol
[MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, MSFT, ...]
;;Varargs version is now a lookup ala clojure.core/get with a not-found
;;value, as opposed to the assoc-like behavior from before.
user> (csv-data  "nope" :not-found)
:not-found
;;assoc works at the column level via add-or-update-column
user> (def added (assoc csv-data "new-column" (repeat (count (first (vals csv-data))) :new-column)))
#'user/added
user> (added "new-column")
#tech.ml.dataset.column<object>[560]
new-column
[:new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, :new-column, ...]

;;dissoc works like dropping columns
user> (def dropped (dissoc csv-data (first (keys csv-data))))
#'user/dropped
user> (keys dropped)
("date" "price")
user> (first (ds/mapseq-reader dropped))
{"date" #object[java.time.LocalDate 0x43e666c9 "2000-01-01"], "price" 39.81}

;;iterator still works...
user> (take 10 (first (iterator-seq (.iterator ^java.lang.Iterable csv-data))))
("MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT")
;;we can use vals to get the same result.
user> (take 10 (first (vals csv-data)))
("MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT" "MSFT")
```

**Comment by cnuernber** [comment]
Working branch here:
https://github.com/techascent/tech.ml.dataset/tree/datasets-as-maps

**Comment by joinr** [comment]
Revised patch fixes printing.

[patch.diff](https://github.com/techascent/tech.ml.dataset/files/4892269/patch.diff.txt)

**Comment by cnuernber** [comment]
Closed in `3.01`.
